modules/container/container.go

```diff
@@ -0,0 +1,219 @@
+	"github.com/containerd/containerd/namespaces"
+	"github.com/containerd/containerd/oci"
+	"github.com/google/shlex"
+	"github.com/threefoldtech/zbus"
+	"github.com/threefoldtech/zosv2/modules"
+	"github.com/threefoldtech/zosv2/modules/stubs"
+)
+
+const (
+	containerdSock = "/run/containerd/containerd.sock"
+)
+
+type containerModule struct {
+	flister    *stubs.FlisterStub
+	containerd string
+}
+
+func New(zcl zbus.Client, containerd string) modules.ContainerModule {
+	if len(containerd) == 0 {
+		containerd = containerdSock
+	}
+
+	return &containerModule{
+		flister:    stubs.NewFlisterStub(zcl),
+		containerd: containerd,
+	}
+}
+
+func getNetworkSpec(network modules.NetworkInfo) oci.SpecOpts {
+	ns := network.Namespace
+	if !path.IsAbs(ns) {
+		// just name
+		ns = path.Join("/var/run/netns", ns)
+	}
+
+	return oci.WithLinuxNamespace(
+		specs.LinuxNamespace{
+			Type: specs.NetworkNamespace,
+			Path: ns,
+		},
+	)
+}
+
+func withHooks(hooks specs.Hooks) oci.SpecOpts {
+	return func(_ context.Context, _ oci.Client, _ *containers.Container, spec *oci.Spec) error {
+		spec.Hooks = &hooks
+		return nil
+	}
+}
+
+// NOTE:
+// THIS IS A WIP Create action and it's not fully implemented atm
+func (c *containerModule) Run(ns string, name string, flist string, tags, env []string, network modules.NetworkInfo,
+	mounts []modules.MountInfo, entrypoint string) (id modules.ContainerID, err error) {
+	// create a new client connected to the default socket path for containerd
+	client, err := containerd.New(c.containerd)
+	if err != nil {
+		return id, err
+	}
+	defer client.Close()
+
+	args, _ := shlex.Split(entrypoint)
+
+	root, err := c.flister.Mount(flist, "")
+	if err != nil {
+		return id, err
+	}
+
+	defer func() {
+		if err != nil {
+			c.flister.Umount(root)
+		}
+	}()
+
+	opts := []oci.SpecOpts{
+		oci.WithDefaultSpecForPlatform("linux/amd64"),
+		oci.WithRootFSPath(root),
+		oci.WithProcessArgs(args...),
+		oci.WithEnv(env),
+
+		// NOTE: the hooks run inside runc namespace
+		// it means that we can't do the unmount of the
+		// root fs from here.
+
+		// withHooks(specs.Hooks{
+		// 	Poststop: []specs.Hook{
+		// 		{
+		// 			Path: "umount",
+		// 			Args: []string{path},
+		// 		},
+		// 	},
+		// }),
+	}
+
+	if len(network.Namespace) != 0 {
+		opts = append(
+			opts,
+			getNetworkSpec(network),
+		)
+	}
+
+	for _, mount := range mounts {
+		opts = append(
+			opts,
+			oci.WithMounts([]specs.Mount{
+				{
+					Destination: mount.Target,
+					Type:        mount.Type,
+					Source:      mount.Source,
+					Options:     mount.Options,
+				},
+			}),
+		)
+	}
+
+	ctx := namespaces.WithNamespace(context.Background(), ns)
+
+	container, err := client.NewContainer(
+		ctx,
+		name,
+		containerd.WithNewSpec(opts...),
+	)
+
+	if err != nil {
+		return id, err
+	}
+
+	defer func() {
+		if err != nil {
+			container.Delete(ctx, containerd.WithSnapshotCleanup)
+		}
+	}()
+
+	logs := path.Join("/var/log", ns)
```

Comment:
Not sure we want to writes logs into `/var/log`
My idea is that a module will have a dedicated root directory under which it can do whatever.
This makes it easy to managed the storage used by each module and avoid spreading files all over the tree

So I would create a log directory under the root directory of the container module and use this instead

Reply:
Yup, makes sense. will change that.